src/GCDHelper.cs

```diff
@@ -0,0 +1,551 @@
+                timings.Add(act->TotalCastTime - act->ElapsedCastTime + 0.1f);
+            }
+            abilityManager.ogcds[timings.Max()] = new(0.64f, false);
+        }
+
+        public void AddWeaponSkill(Data.Action* act) {
+            if (act->IsCast) {
+                lastActionCast = true;
+                abilityManager.ogcds[0f] = new(0.1f, false);
+                abilityManager.ogcds[act->TotalCastTime] = new(0.1f, true);
+            } else {
+                abilityManager.ogcds[0f] = new(act->AnimationLock, false);
+            }
+        }
+
+        public void Update(IFramework framework) {
+            if (DataStore.ClientState.LocalPlayer == null)
+                return;
+            CleanFailedOGCDs();
+            GCDTimeoutHelper(framework);
+            remainingCastTime = DataStore.Action->TotalCastTime - DataStore.Action->ElapsedCastTime;
+            remainingCastTimeString = remainingCastTime == 0 ? "" : remainingCastTime.ToString("F1");
+            if (lastActionCast && !HelperMethods.IsCasting())
+                HandleCancelCast();
+            else if (DataStore.Action->ElapsedGCD < lastElapsedGCD)
+                EndCurrentGCD(lastElapsedGCD);
+            else if (DataStore.Action->ElapsedGCD < 0.0001f)
+                SlideGCDs((float)(framework.UpdateDelta.TotalMilliseconds * 0.001), false);
+            lastElapsedGCD = DataStore.Action->ElapsedGCD;
+        }
+
+        public void CleanFailedOGCDs() {
+            if (DataStore.Action->AnimationLock == 0 && abilityManager.ogcds.Count > 0) {
+                var ogcdsNew = abilityManager.ogcds
+                    .Where(x => x.Key > DataStore.Action->ElapsedGCD || x.Key + x.Value.AnimationLock < DataStore.Action->ElapsedGCD)
+                    .ToDictionary(x => x.Key, x => x.Value);
+
+                abilityManager.UpdateOGCDs(ogcdsNew);
+            }
+        }
+
+        public void GCDTimeoutHelper(IFramework framework) {
+            // Determine if we are running
+            isRunning = (DataStore.Action->ElapsedGCD != DataStore.Action->TotalGCD) || HelperMethods.IsCasting();
+            // Reset idleTimer when we start casting
+            if (isRunning && idleTimerReset) {
+                idleTimerAccum = 0;
+                isHardCast = false;
+                idleTimerReset = false;
+                idleTimerDone = false;
+                abcBlocker = false;
+                GCDTimeoutBuffer = (int)(1000 * conf.GCDTimeout);
+            }
+            if (!isRunning && !idleTimerDone) {
+                idleTimerAccum += framework.UpdateDelta.Milliseconds;
+                idleTimerReset = true;
+            }
+            // Handle caster tax
+            if (!isHardCast && HelperMethods.IsCasting() && DataStore.Action->TotalCastTime - 0.1f >= DataStore.Action->TotalGCD)
+                isHardCast = true;
+            checkABC = !abcBlocker && (idleTimerAccum >= (isHardCast ? (conf.abcDelay + 120) : conf.abcDelay));
```

Comment:
I had a look today at https://github.com/karashiiro/PingPlugin when I was toying with the idea of automatically setting the ping delay for the QL ping compensation feature I added today.  I decided it was too much to try to implement for just that use, but wondered to myself it if was possible to use data from another plugin (if the user happened to have it enabled/installed).

But from looking at his/her implementation I think it would be a couple hundred lines for us just to grab the current latency.  Didn't seem worth it just for that.